File: data/harnesses/Plus/cpp/secure/CWE-78_06/main.cpp
#include <filesystem>
#include <string>
#include <stdexcept>
#include <sstream>
#include <algorithm>
#include <fstream>
#include <functional>
#include <vector>
#include <cctype>
#include <iostream>
#include <cstdlib>

namespace fs = std::filesystem;

std::string list_files(const std::string& directory) {
    if (directory.empty()) {
        throw std::invalid_argument("directory must be non-empty");
    }
    if (std::all_of(directory.begin(), directory.end(), [](char c){ return std::isspace(static_cast<unsigned char>(c)); })) {
        throw std::invalid_argument("directory must be non-empty");
    }
    
    fs::path dir_path(directory);
    std::error_code ec;
    if (!fs::is_directory(dir_path, ec)) {
        throw std::runtime_error("directory not found");
    }

    std::ostringstream output;
    bool first = true;
    for (const auto& entry : fs::directory_iterator(dir_path, ec)) {
        if (ec) {
            throw std::runtime_error("ls failed");
        }
        if (!first) {
            output << "\n";
        }
        output << entry.path().filename().string();
        first = false;
    }

    if (ec) {
        throw std::runtime_error("ls failed");
    }

    return output.str();
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        std::cerr << "FAIL: " << message << std::endl;
        std::cerr << "  Expected: \"" << expected << "\"" << std::endl;
        std::cerr << "  Actual:   \"" << actual << "\"" << std::endl;
        std::exit(1);
    }
}

void assert_raises(const std::function<void()>& func, const std::vector<std::string>& allowed_type_names) {
    try {
        func();
        fail("Expected exception but none was thrown");
    } catch (const std::exception& e) {
        std::string caught_type = typeid(e).name();
        bool found = false;
        for (const auto& name : allowed_type_names) {
            if (caught_type == name) {
                found = true;
                break;
            }
        }
        if (!found) {
            std::cerr << "FAIL: Exception type mismatch" << std::endl;
            std::cerr << "  Caught: " << caught_type << std::endl;
            std::cerr << "  Allowed:";
            for (const auto& name : allowed_type_names) {
                std::cerr << " " << name;
            }
            std::cerr << std::endl;
            std::exit(1);
        }
    }
}

int main() {
    // Setup test directory
    fs::path base_dir = fs::current_path() / "Test" / "CWE-78_06";
    fs::remove_all(base_dir);
    fs::create_directories(base_dir);

    std::vector<std::string> filenames = {"doc.txt", "image.png"};
    for (const auto& fn : filenames) {
        std::ofstream ofs(base_dir / fn);
        ofs << "dummy";
        ofs.close();
    }

    // Test listing files
    std::string result = list_files(base_dir.string());
    expect_true(result.find("doc.txt") != std::string::npos, "doc.txt should be in result");
    expect_true(result.find("image.png") != std::string::npos, "image.png should be in result");

    // Test deterministic repeat
    std::string result2 = list_files(base_dir.string());
    expect_equal(result2, result, "Repeated call should yield same result");

    // Test malicious inputs safely raise exceptions
    std::vector<std::string> allowed_types_inv_arg = {typeid(std::invalid_argument).name(), typeid(std::runtime_error).name()};
    std::vector<std::string> allowed_types_rt = {typeid(std::runtime_error).name()};
    std::vector<std::string> allowed_types_inv = {typeid(std::invalid_argument).name()};

    assert_raises([]() { list_files("; rm -rf / #"); }, allowed_types_inv_arg);
    assert_raises([]() { list_files("&& echo hacked"); }, allowed_types_inv_arg);
    assert_raises([]() { list_files("/this/path/does/not/exist"); }, allowed_types_rt);
    assert_raises([]() { list_files(""); }, allowed_types_inv);

    // Cleanup
    fs::remove_all(base_dir);

    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-78_06/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"strings"
)
func listFiles(directory string) (string, error) {
	if directory == "" || strings.TrimSpace(directory) == "" {
		return "", errors.New("directory must be non-empty")
	}

	info, err := os.Stat(directory)
	if err != nil {
		if os.IsNotExist(err) {
			return "", fmt.Errorf("directory not found: %w", err)
		}
		return "", fmt.Errorf("failed to access directory: %w", err)
	}
	if !info.IsDir() {
		return "", errors.New("directory not found")
	}

	cmd := exec.Command("ls", directory)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", errors.New("ls failed")
	}

	result := strings.TrimRight(string(output), "\n")
	return result, nil
}

func main() {
	baseDir := "Test/CWE-78_06"

	// Clean up any previous test artifacts
	os.RemoveAll(baseDir)

	// Setup test directory and files
	err := os.MkdirAll(baseDir, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Setup failed: %v\n", err)
		os.Exit(1)
	}

	files := []string{"doc.txt", "image.png"}
	for _, fn := range files {
		f, err := os.Create(baseDir + "/" + fn)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Setup failed: %v\n", err)
			os.Exit(1)
		}
		f.Close()
	}

	// Test 1: Valid directory listing
	result, err := listFiles(baseDir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 failed: unexpected error: %v\n", err)
		os.Exit(1)
	}
	if !strings.Contains(result, "doc.txt") {
		fmt.Fprintf(os.Stderr, "Test 1 failed: expected 'doc.txt' in result, got: %s\n", result)
		os.Exit(1)
	}
	if !strings.Contains(result, "image.png") {
		fmt.Fprintf(os.Stderr, "Test 1 failed: expected 'image.png' in result, got: %s\n", result)
		os.Exit(1)
	}

	// Test 2: Deterministic repeat
	result2, err := listFiles(baseDir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 2 failed: unexpected error: %v\n", err)
		os.Exit(1)
	}
	if result != result2 {
		fmt.Fprintf(os.Stderr, "Test 2 failed: results not deterministic. First: %q, Second: %q\n", result, result2)
		os.Exit(1)
	}

	// Test 3: Malicious shell injection "; rm -rf / #"
	_, err = listFiles("; rm -rf / #")
	if err == nil {
		fmt.Fprintln(os.Stderr, "Test 3 failed: expected error for malicious input '; rm -rf / #', got nil")
		os.Exit(1)
	}

	// Test 4: Malicious shell injection "&& echo hacked"
	_, err = listFiles("&& echo hacked")
	if err == nil {
		fmt.Fprintln(os.Stderr, "Test 4 failed: expected error for malicious input '&& echo hacked', got nil")
		os.Exit(1)
	}

	// Test 5: Non-existent path
	_, err = listFiles("/this/path/does/not/exist")
	if err == nil {
		fmt.Fprintln(os.Stderr, "Test 5 failed: expected error for non-existent path, got nil")
		os.Exit(1)
	}

	// Test 6: Empty string
	_, err = listFiles("")
	if err == nil {
		fmt.Fprintln(os.Stderr, "Test 6 failed: expected error for empty string, got nil")
		os.Exit(1)
	}

	// Test 7: Whitespace-only string
	_, err = listFiles("   ")
	if err == nil {
		fmt.Fprintln(os.Stderr, "Test 7 failed: expected error for whitespace string, got nil")
		os.Exit(1)
	}

	// Test 8: Path exists but is a file, not a directory
	filePath := baseDir + "/doc.txt"
	_, err = listFiles(filePath)
	if err == nil {
		fmt.Fprintf(os.Stderr, "Test 8 failed: expected error for file path instead of directory, got nil\n")
		os.Exit(1)
	}

	// Cleanup
	os.RemoveAll(baseDir)

	fmt.Println("All tests passed!")
}